contribute/design/proposal-volume-snapshots-clone.md

```diff
@@ -0,0 +1,97 @@
+
+- User wants to share(read-only) the data stored on openebs volume with multiple clients. 
+  Possible solution:
+  * Once the data that needs to be shared is available on the OpenEBS volume, create a snapshot. 
+  * Create a Clone of the volume from the snapshot. And provide ReadOnlyMany access to the cloned volume. 
+
+- User is running a Database on the OpenEBS Volume. User wants to try out some schema upgrades to the Database which will determine if the new changes need to be applied or the database needs to be restored back to its original state. 
+  Possible solution:
+  * Create a snapshot of the OpenEBS Volume associated with the database
+  * Create a Clone of the volume from the snapshot and Run the database onthe cloned volume.
+  * Apply the schema changes on the new database and validate the changes.
+  * Commit the validated schema changes and the upgrade sql (data migration) scripts on the original database
+  * Deleted the cloned volume and the snapshot that was used to run the tests.
+
+## Usecases (Non-Goals)
+- Uses wishes to setup a process to protect the data stored on openebs volumes against system (kubernetes nodes or openebs storage) failures.
+  Possible solution:
+  * Snapshot the OpenEBS Volumes at regular intervals like daily or weekly. Transfer the data from the snapshot to a remote location - external from the current system (kubernetes nodes and storage). The remote locations could be an S3 store or another OpenEBS Volume in a different location.
+  * In the event of an system failure, restore the data from the backup location - s3 or remote openebs volume.
+
+
+## Desgin
+
+The usecases related to protecting data against application failures and recovering data from local snapshots can be achieved if the following primitives(API) are supported by the OpenEBS Volumes:
+- Create Snapshot
+- Delete Snapshot
+- List Snapshot
+- Mount Snapshot - same as Clone Volume with ReadOnly Option
+- Unmount Snapshot - same as Delete Clone Volume.
+- Clone Volume
+- Delete Clone Volume
+- Revert Volume State to a previous Snapshot
+
+Along wih the OpenEBS Volume Dynamic Provisioner, OpenEBS Snapshot Provisioner will be launched during the openebs install. The Snapshot Provisioner is an extension/plugin of https://github.com/kubernetes-incubator/external-storage/tree/master/snapshot
+
+### Create Snapshot
+* Admin uses the kubectl volume snapshot yaml to create a snapshot on a given PV/PVC. 
+* Kubernets will forward the snapshot create request to the OpenEBS Snapshot provisioner, while will do the following:
+  - Identify the cStorService or the JivaService associated with the PV
+  - If Jiva, call the REST API of Jiva Controller to create a new snapshot
+  - If cStor, call the REST API of cStorVolumeMgmt(side-car?) to create a new snapshot
```

Comment:
I remember three-step approach for taking snapshots. @gila .. would you think its still a better approach?

Reply:
Updated with the flow we discussed using the side-cars and CRDs.